### PR TITLE
[WIP] Prevent Prepare from evaluating SQL expressions.

### LINKED ISSUE #5340: sql: Prepare incorrectly evaluates expressions
During Prepare expressions should be left unevaluated, especially built-in functions marked "impure".

Found in #5333.

### PR DESCRIPTION
**Pull request by knz**
Fixes #5340.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5357)

### CODE REVIEW

File: sql/pgwire_test.go [hunk 1]
```diff
@@ -446,7 +461,7 @@ func TestPGPreparedQuery(t *testing.T) {
 		}
 	}
 
-	initStmt := `CREATE DATABASE d; CREATE TABLE d.t (a INT); INSERT INTO d.t VALUES (10),(11); CREATE TABLE d.ts (a TIMESTAMP, b DATE);`
+	initStmt := `CREATE DATABASE d; CREATE TABLE d.t (a INT); INSERT INTO d.t VALUES (10),(11); CREATE TABLE d.ts (a TIMESTAMP, b DATE); CREATE TABLE d.tsd (a TIMESTAMP DEFAULT NOW(), b INT);`
```
Comment (petermattis) [inline sql/pgwire_test.go]: No reason for this to be one long line. Let's add some newlines after every statement for readability.

File: sql/values.go [hunk 2]
```diff
@@ -65,6 +65,15 @@ func (p *planner) ValuesClause(n *parser.ValuesClause) (planNode, *roachpb.Error
...
 			}
 		}
+		// TODO(knz): fix #5340
+		// We should stop here if working for planner.Prepare.  However
+		// for now we cannot stop here because planner.Insert() doesn't
+		// know how to deal with a result with only column
+		// names/types. The following simply doesn't work:
+		//	if p.prepareOnly {
+		//		continue
+		//	}
```
Comment (petermattis) [inline sql/values.go]: I think we need to fix `planner.Insert` to use `columns[i].Typ` when `prepareOnly` is true.

File: sql/pgwire_test.go (hunk 1 above)
Comment (knz) [inline sql/pgwire_test.go]: done

**Comment by RaduBerinde** [comment]
LGTM. There is a lot of flakiness in the prepare implementation in general. It would be nice to have a testing framework that takes a set of statements and tries to make each and every possible Datum in the AST a prepared argument (in various combinations).

---

Reviewed 9 of 12 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/plan.go, line 57 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5357#-KD_vPDuKr6eqfW8xya7:-KD_vPDvH9CEE7e3rmH4:1934208894) ([raw file](https://github.com/cockroachdb/cockroach/blob/1257b3ef73a429e860134e677361689702d5ec9d/sql/plan.go#L57)):*
I know you're just moving this comment but I wanted to note something - we can't really have a complete/correct plan during prepare. We can't properly select an index with unknown values in the filter, we may not be able to propagate limits, etc. When those values are filled in, things can change drastically - even the structure of the plan itself can change, e.g. we may not need a `sortNode` anymore.  Instead, we should avoid doing unnecessary work during prepare (e.g. why try to select an index at all; we may not even need to generate a "full" plan).

---

*[sql/parser/builtins.go, line 554 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5357#-KD_y773KN3fKIAqtHCX:-KD_y773KN3fKIAqtHCY:1540091434) ([raw file](https://github.com/cockroachdb/cockroach/blob/1257b3ef73a429e860134e677361689702d5ec9d/sql/parser/builtins.go#L554)):*
Can you briefly mention this change in the commit message?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5357#-:-KD_xMpfg2ybs8azXFso:2144158991)*

**Comment by tamird** [comment]
Reviewed 9 of 12 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[sql/pgwire_test.go, line 389 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5357#-KDdncTT4jAzr_rVqtoY:-KDdncTT4jAzr_rVqtoZ:-733417483) ([raw file](https://github.com/cockroachdb/cockroach/blob/1257b3ef73a429e860134e677361689702d5ec9d/sql/pgwire_test.go#L389)):*
how come these tests need to be commented out? (the existing ones, not the new ones)

is it because of the `impure` mark on `current_date`?

---

*[sql/pgwire_test.go, line 400 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5357#-KDdo26IcKAB-TtFTzOY:-KDdo26IcKAB-TtFTzOZ:-2006093357) ([raw file](https://github.com/cockroachdb/cockroach/blob/1257b3ef73a429e860134e677361689702d5ec9d/sql/pgwire_test.go#L400)):*
how come you're changing this test? looks like the new thing is testing something different from what was being tested before.

---

*[sql/plan.go, line 184 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5357#-KDdnBYnQQ390N8LN9Ba:-KDdnBYnQQ390N8LN9Bb:-642914758) ([raw file](https://github.com/cockroachdb/cockroach/blob/1257b3ef73a429e860134e677361689702d5ec9d/sql/plan.go#L184)):*
do you need to `defer func() { p.prepareOnly = false }()` here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5357)*

**Comment by knz** [comment]
Superseded by #5903.

**Pull request closed without merging**